// @fluentui/utilities/lib-commonjs/aria.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var aria_1 = require("./aria");
describe('aria utils tests', function () {
    describe('mergeAriaAttributeValues tests', function () {
        var mergeTestCases = [
            {
                description: 'returns undefined when given no args',
                cases: [
                    {
                        args: [],
                        expected: undefined,
                    },
                ],
            },
            {
                description: 'returns undefined when given undefined and empty args',
                cases: [
                    {
                        args: [undefined],
                        expected: undefined,
                    },
                    {
                        args: [undefined, undefined],
                        expected: undefined,
                    },
                    {
                        args: [''],
                        expected: undefined,
                    },
                    {
                        args: [undefined, ''],
                        expected: undefined,
                    },
                ],
            },
            {
                description: 'returns arg when given one valid arg',
                cases: [
                    {
                        args: ['arg1'],
                        expected: 'arg1',
                    },
                    {
                        args: ['arg1', undefined],
                        expected: 'arg1',
                    },
                    {
                        args: [undefined, 'arg1', undefined],
                        expected: 'arg1',
                    },
                    {
                        args: ['', 'arg1', ''],
                        expected: 'arg1',
                    },
                ],
            },
            {
                description: 'returns merged args when given multiple valid args',
                cases: [
                    {
                        args: ['arg1', 'arg2'],
                        expected: 'arg1 arg2',
                    },
                    {
                        args: ['arg1', undefined],
                        expected: 'arg1',
                    },
                    {
                        args: [undefined, 'arg1', undefined],
                        expected: 'arg1',
                    },
                    {
                        args: ['', 'arg1', ''],
                        expected: 'arg1',
                    },
                    {
                        args: ['', 'arg1', 'arg2 '],
                        expected: 'arg1 arg2',
                    },
                    {
                        args: ['', ''],
                        expected: undefined,
                    },
                    {
                        args: [' ', ' '],
                        expected: undefined,
                    },
                ],
            },
        ];
        mergeTestCases.forEach(function (test) {
            test.cases.forEach(function (testCase, index) {
                it(test.description + ', case #' + index, function () {
                    var merged = aria_1.mergeAriaAttributeValues.apply(void 0, testCase.args);
                    expect(merged).toEqual(testCase.expected);
                });
            });
        });
    });
});
