diff --git a/node_modules/@griffel/react/utils/isInsideComponent.esm.js b/node_modules/@griffel/react/utils/isInsideComponent.esm.js
index 79cead2..de84720 100644
--- a/node_modules/@griffel/react/utils/isInsideComponent.esm.js
+++ b/node_modules/@griffel/react/utils/isInsideComponent.esm.js
@@ -5,7 +5,7 @@ function isInsideComponent() {
   // https://github.com/facebook/react/blob/42f15b324f50d0fd98322c21646ac3013e30344a/packages/react/src/ReactHooks.js#L26-L36
   try {
     // @ts-expect-error "SECRET_INTERNALS" are not typed
-    const dispatcher = React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentDispatcher.current;
+    const dispatcher = React.__CLIENT_INTERNALS_DO_NOT_USE_OR_WARN_USERS_THEY_CANNOT_UPGRADE.ReactCurrentDispatcher.current;
     // Before any React component was rendered "dispatcher" will be "null"
     if (dispatcher === null || dispatcher === undefined) {
       return false;
